### PR TITLE
feat: Implement DuckDB storage for diagram runtime data in ds-ext

### PR DESCRIPTION
**Pull request by stone-lyl**
I use the SQLTools extension to view the database in VSCode, but the experience with this extension isn't great 😂 

image

image

image

Use DuckDB to store runtime data for diagrams.

https://github.com/user-attachments/assets/b2013f3a-5152-48be-89f7-7aeb870cefaa

### CODE REVIEW

File: packages/core/src/InputObserverController.ts [hunk 1]
```diff
@@ -56,21 +56,23 @@ export class InputObserverController {
...
   }
 
-  getDataFromStorage({
+  // todo: get data from database set limit and offset
+  async getDataFromStorage({
     linkId,
     limit = 100,
     offset = 0
-  }: GetDataFromStorage): Record<LinkId, ItemValue[]> {
+  }: GetDataFromStorage): Promise<Record<LinkId, ItemValue[]>> {
```
Comment (ajthinking) [inline packages/core/src/InputObserverController.ts:73]: Consider an intermediate type, perhaps `type LinkItems = Record<LinkId, ItemValue[]>`

File: packages/ds-ext/src/duckDBStorage.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        createTime TIMESTAMP,
+        updateTime TIMESTAMP
+      );
+      CREATE INDEX idx_linkId ON linkItems (linkId);
+    
+      CREATE TABLE IF NOT EXISTS nodes (
+        nodeId TEXT PRIMARY KEY,
+        status TEXT,
+        createTime TIMESTAMP,
+        updateTime TIMESTAMP
```
Comment (ajthinking) [inline packages/ds-ext/src/duckDBStorage.ts:35]: Do we have a use case for updateTime or is it just for a uniform/standard setup to have it?

File: packages/ds-ext/src/duckDBStorage.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+      );
+    `);
+  }
+
+  async close() {
+    if (this.db) {
+      await this.db.close();
```
Comment (ajthinking) [inline packages/ds-ext/src/duckDBStorage.ts:42]: Can be written as 
```ts
await this.db?.close();
```

File: packages/ds-ext/src/duckDBStorage.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    const data = await this.db?.all('SELECT item FROM linkItems WHERE linkId = ?', linkId);
+    if (!data || data.length === 0) {
+      return undefined;
+    }
+    const result = data.map(row => JSON.parse(row.item));
+    return result;
+  }
+
+  async setLinkItems(linkId: string, items: Record<string, any>[]): Promise<void> {
+    await this.db?.all('DELETE FROM linkItems WHERE linkId = ?', linkId);
```
Comment (ajthinking) [inline packages/ds-ext/src/duckDBStorage.ts:70]: Is there a reason to prefer DELETE + reinsert over an UPDATE query?

File: packages/ds-ext/src/extension.ts [hunk 5]
```diff
@@ -3,18 +3,39 @@ import { DiagramEditorProvider } from './DiagramEditorProvider';
...
   context.subscriptions.push(
     vscode.window.registerCustomEditorProvider(
       'ds-ext.diagramEditor',
-      new DiagramEditorProvider(context)
+      diagramEditorProvider
     )
   );
+
+  console.log('ds-ext is installed at ', context.extensionPath);
+  // run yarn install in the extension folder
```
Comment (ajthinking) [inline packages/ds-ext/src/extension.ts:25]: Why is this needed, please explain this a bit 🙏 Seems a bit quirky to install dependencies at activation-time?

File: packages/ds-ext/src/DiagramEditorProvider.ts [hunk 6]
```diff
@@ -35,9 +37,14 @@ export class DiagramEditorProvider implements vscode.CustomEditorProvider<Diagra
...
   }
 
+  dispose(): void {
+    this.duckDBStorage.close();
+  }
+
   constructor(private readonly context: vscode.ExtensionContext) {
-    const storage = new DiagramObserverStorage();
-    this.inputObserverController = new InputObserverController(storage);
+    const dbPath = path.join(vscode.workspace.workspaceFolders![0].uri.fsPath, 'diagram.db');
```
Comment (ajthinking) [inline packages/ds-ext/src/DiagramEditorProvider.ts:45]: I suggest we move this to `./datastory/executions.db` to indicate it being temporary/git ignorable and describe the content a tiny bit clearer

File: packages/ds-ext/src/duckDBStorage.ts (hunk 2 above)
Comment (stone-lyl) [reply packages/ds-ext/src/duckDBStorage.ts:35]: The `updateTime` field is primarily used for a uniform/standard setup. 
However, it can also serve as a data observation tool during development. For instance, in the `node` table, we can track the time taken for a node to progress from the initial state to available status and then to complete status using `createTime` and `updateTime`.

File: packages/ds-ext/src/duckDBStorage.ts (hunk 4 above)
Comment (stone-lyl) [reply packages/ds-ext/src/duckDBStorage.ts:70]: During a re-run, I clear all item data corresponding to the `linkId` in the current diagram and insert the item data for the `linkId` from the latest execution run.

If we need to retain the results of each execution going forward, we won't need this reset operation anymore.

File: packages/ds-ext/src/extension.ts (hunk 5 above)
Comment (stone-lyl) [reply packages/ds-ext/src/extension.ts:25]: The reason is that the DuckDB binary is not compatible across all platforms (macOS, Windows, Linux), so users need to install DuckDB locally (this approach is based on practices from DuckDB SQL and other VSCode extensions).
